packages/tools/client-debugger/client-debugger/src/data-visualization/DataVisualization.ts

```diff
@@ -0,0 +1,444 @@
+	extends TypedEventEmitter<DataVisualizerEvents>
+	implements IDisposable
+{
+	/**
+	 * {@inheritDoc IFluidClientDebugger.containerData}
+	 */
+	private readonly rootData: Record<string, IFluidLoadable>;
+
+	/**
+	 * Policy object for visualizing different kinds of shared objects.
+	 */
+	private readonly visualizers: SharedObjectVisualizers;
+
+	/**
+	 * Map of registered {@link VisualizerNode}s, keyed by their corresponding {@link FluidObjectId}.
+	 *
+	 * @privateRemarks TODO: Dependency tracking so we don't leak memory.
+	 */
+	private readonly visualizerNodes: Map<FluidObjectId, VisualizerNode>;
+
+	/**
+	 * Private {@link VisualizerNode.disposed} tracking.
+	 */
+	private _disposed: boolean;
+
+	/**
+	 * Handler for a visualizer node's "update" event.
+	 * Bubbles up the event to graph subscribers.
+	 */
+	private readonly onVisualUpdateHandler = (visualTree: FluidObjectNode): boolean => {
+		this.emitVisualUpdate(visualTree);
+		return true;
+	};
+
+	public constructor(
+		rootData: Record<string, IFluidLoadable>,
+		visualizers: SharedObjectVisualizers,
+	) {
+		super();
+
+		this.rootData = rootData;
+		this.visualizers = visualizers;
+
+		this.visualizerNodes = new Map<FluidObjectId, VisualizerNode>();
+
+		this._disposed = false;
+	}
+
+	/**
+	 * {@inheritDoc IDisposable.disposed}
+	 */
+	public get disposed(): boolean {
+		return this._disposed;
+	}
+
+	/**
+	 * Emits a visual tree update from one of the registered visualizer nodes.
+	 */
+	private emitVisualUpdate(visualTree: FluidObjectNode): void {
+		this.emit("update", visualTree);
+	}
+
+	/**
+	 * Generates and returns visual descriptions ({@link FluidHandleNode}s) for each of the specified
+	 * {@link DataVisualizerGraph.rootData | root shared objects}.
+	 */
+	public async renderRootHandles(): Promise<FluidHandleNode[]> {
+		// Rendering the root entries amounts to initializing visualizer nodes for each of them, and returning
+		// a list of handle nodes. Consumers can request data for each of these handles as needed.
+		const rootDataEntries = Object.entries(this.rootData);
+		return Promise.all(
+			rootDataEntries.map(async ([key, value]) => {
+				const fluidObjectId = await this.registerVisualizerForHandle(value.handle, key);
+				return createHandleNode(fluidObjectId, key);
+			}),
+		);
+	}
+
+	/**
+	 * Generates and returns a visual description of the specified Fluid object if it exists in the graph.
+	 * If no such object exists in the graph, returns `undefined`.
+	 */
+	public async render(fluidObjectId: FluidObjectId): Promise<FluidObjectNode | undefined> {
+		if (!this.visualizerNodes.has(fluidObjectId)) {
+			// We don't have anything registered for the requested Fluid object.
+			// This could indicate a stale data request from an external consumer, or could indicate a bug.
+			return undefined;
+		}
+
+		// Checked above.
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const visualizerNode = this.visualizerNodes.get(fluidObjectId)!;
```

Comment:
```suggestion
		return visualizerNode.render();
```

Reply:
It might be better to make request to `visualizerNodes` only once

Reply:
Cleaned up in [6ed50a4](https://github.com/microsoft/FluidFramework/pull/14770/commits/6ed50a4b0d3e302b6d545630564e52c48ef35d42)